### ISSUE TITLE
import of matplotlib causes error on Mac OS X (depending on backend)

### ISSUE THREAD

**Issue by lukasheinrich** [scikit-optimize/scikit-optimize#637]
Since skopt import matplotlib for its plotting, it runs into some known issues with Mac OSX
```
    from matplotlib.backends import _macosx
RuntimeError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework. See the Python documentation for more information on installing Python as a framework on Mac OS X. Please either reinstall Python as a framework, or try one of the other backends. If you are using (Ana)Conda please install python.app and replace the use of 'python' with 'pythonw'. See 'Working with Matplotlib on OSX' in the Matplotlib FAQ for more information.
```

this can be fixed by 

```
import matplotlib
matplotlib.use('PS')
```

before importing `matplotlib.pyplot`

but it would be nice if skopt could import such that it ignores those plotting-related issues for cases when one is not interested in plotting

**Comment by betatim** [comment]
Do you know how other projects handle this? I think the reason matplotlib gets imported as soon as you import skopt is because skopt imports the plotting things at the top level. Would be nice to delay that to fix this error and speed up initial skopt imports.

---

How do you get the error? I am on mac and haven't ever seen it :(

**Comment by lukasheinrich** [comment]
@betatim I think it's related to whether you use the system native python or not (including usage of virtualenvs.... e.g. for me, I get this behaviour

```
$> pyenv shell system
$> python
>>> import matplotlib.pyplot as plt #works
```

```
$> pyenv shell system
$> virtualenv venv
$> source venv/bin/activate
$> pip install matplotlib
$> python
>>> import matplotlib.pyplot as plt #fails
```

```
$> pyenv shell 3.5.1
$> python
>>> import matplotlib.pyplot as plt #fails
```

I think it would be reasonable to have sth like

```
try:
   import matplotlib.pyplot as plt
except RuntimeError as e:
   if 'Python is not installed as a framework.' in e.message:
     warnings.warn(.. some warning about disabled backend...)
```

you could also just handle any ImportError or RuntimeError resulting from the plotting imports and handle them in a non-raising way

**Comment by ledona** [comment]
I was able to work around the problem by forcing matplotlib to use TCL/TK and reinstalling python with tcl. This solved the problem both when running in a virtual env and in the native python environment.

I'm using homebrew so the following is what I did.

```
echo "backend: TkAgg" >> ~/.matplotlib/matplotlibrc
brew uninstall python3
brew install python3 --with-tcl-tk
```

**Comment by lukasheinrich** [comment]
Yes, a number of workarounds exist (e.g. 'PS' backend in OP), but I was hoping that skopt would only import matplotlib when I actually want to do some plotting and not earlier.

**Comment by betatim** [comment]
I think in general fixing your dependencies to be properly installed is up to the user. scikit-optimize shouldn't try and fix it for you or do weird stuff to help users who can't install things.

However delaying the import of matplotlib until it is actually needed is a useful feature to have because it would speed up the initial skopt import and allow us to make it an optional dependency (for example if you are running on a headless node). I think this approach would also solve the original problem described here.

If anyone knows a nice pattern used by other libraries to support optional dependencies please post them. So far all I could find was the pandas approach of importing in a class and several extra modules that delay imports. A widely used pattern would be nice to reduce the chances of weird side effects :-/

**Comment by carlosdanielcsantos** [comment]
@betatim 

There are good reasons for why scientific libraries don't usually have matplotlib as a direct dependency: it is very platform dependent and it is usually not essential for what these libraries are trying to accomplish. Why should an optimization library depend so heavily on a visualization library?

I agree that you shouldn't try to fix dependencies for the user and I would go even further: you shouldn't even install it for them. 

Take pandas as example: they have the plot() method on dataframes and series that call matplotlib.pyplot.plot() but they only import it at the time of the call and matplotlib isn't even a dependency of pandas! And the reason is very simple: pandas is a data manipulation library, the visualization part is just a kind of add-on. Leaving matplotlib out by default instead of pushing it into users that don't even need it seems like a reasonable decision to make.

I'm trying to solve a similar problem to the one described. We have a library that will depend on scikit-optimize but we are having problems because we can't depend on matplotlib platform sensitiveness and, sincerely, we just need the optimizing functions.

Is a change like this something you are willing to accept? We can propose a PR for this if you are!

Thanks for the great library you have here!

**Comment by lukasheinrich** [comment]
I agree with @carlosdanielcsantos though I think it might be reasonable to install the libraries as part of an extra, i.e. `pip install skopt[plotting]`.  (I think this is what you asked for @betatim, it's a way specify optional dependencies  in setuptools).

**Comment by betatim** [comment]
This is exactly what I implemented in #639 :) It has been sitting there unmerged because I didn't know off the top of my head how to add a test for the behaviour. Maybe we can merge it without a test and someone contributes that test in a new PR.

**Comment by carlosdanielcsantos** [comment]
I don't think there's a downside of not importing matplotlib. The only thing we have to guarantee is that all the current tests pass (we need to install with the option "plots" when testing, but that is being done in the PR). If we want to test the behavior of not installing matplotlib we just have to do an assertRaises on `import matplotlib` when scikit-optimize is installed without the "plots" option.

I deferred import to use time just because this is how other well established libraries do it. I sincerely can't see much difference between both approaches (between importing at use time or when importing plots submodule).
